online/src/main/scala/ai/chronon/online/FetcherBase.scala

```diff
@@ -309,14 +305,85 @@ class FetcherBase(kvStore: KVStore,
                   ex.printStackTrace()
                   throw ex
               }
+              logger.info(s"[test] groupByServingInfo ${groupByServingInfo}")
+              logger.info(s"[test] groupByServingInfo ${groupByServingInfo.groupBy.derivationsScala}")
+              if (groupByServingInfo.groupBy.hasDerivations) {
+                logger.info("[test] groupBy has derivations")
+                val keySchema = groupByServingInfo.keyCodec.chrononSchema.asInstanceOf[StructType]
+                val baseValueSchema = if (groupByServingInfo.groupBy.aggregations == null) {
+                  groupByServingInfo.selectedChrononSchema
+                } else {
+                  groupByServingInfo.outputChrononSchema
+                }
+                logger.info(s"[test] baseValueSchema ${baseValueSchema}")
+                logger.info(s"[test] keySchema ${keySchema}")
+                logger.info(s"[test] baseMap ${groupByResponse}")
+                constructGroupByResponseWithDerivation(
+                  groupByServingInfo,
+                  request,
+                  keySchema,
+                  baseValueSchema,
+                  groupByResponse
+                )
+              } else {
+                logger.info("[test] groupBy has no derivations")
+                groupByResponse
+              }
             }
-
             Response(request, responseMapTry)
         }.toList
         responses
       }
   }
 
+  private def constructGroupByResponseWithDerivation(
+      groupByServingInfo: GroupByServingInfo,
+      request: Request,
+      keySchema: StructType,
+      baseValueSchema: StructType,
+      baseMap: Map[String, AnyRef]
+  ): Map[String, AnyRef] = {
+    val requestTs = request.atMillis.getOrElse(System.currentTimeMillis())
+    val requestDs = TsUtils.toStr(requestTs).substring(0, 10)
+    // used for derivation based on ts/ds
+    val tsDsMap: Map[String, AnyRef] =
+      Map("ts" -> (requestTs).asInstanceOf[AnyRef], "ds" -> (requestDs).asInstanceOf[AnyRef])
+    val timeFields: Array[StructField] = Array(
+      StructField("ts", LongType),
+      StructField("ds", StringType)
+    )
+    val conf = groupByServingInfo.groupBy
+    def deriveFunc: (Map[String, Any], Map[String, Any]) => Map[String, Any] = {
```

Comment:
running this function is expensive - we we need to cache the result of the derive function. You can follow how it is done in the join derivation case.

Reply:
I notice that this code is repeated - feel free to refactor to re-use both for join derivations and groupBy derivations.